src/jvmMain/kotlin/com/sdercolin/vlabeler/video/VideoState.kt

```diff
@@ -0,0 +1,73 @@
+package com.sdercolin.vlabeler.video
+
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.awt.ComposeWindow
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import com.sdercolin.vlabeler.audio.PlayerState
+import com.sdercolin.vlabeler.util.toMillisecond
+import java.awt.Container
+
+class VideoState(val debug: Boolean = true) {
+    var width: Dp by mutableStateOf(RememberWidth)
+    var height: Dp = RememberHeight
+
+    var window: ComposeWindow? = null
+    lateinit var container: Container
+
+    var playerState: PlayerState? = null
+    var videoPath: String = ""
+    var currentSampleRate = 0f
+    var embeddedMode: Boolean by mutableStateOf(true)
+
+    val miniVideo = MiniVideo()
+    var afterPopup: Boolean = false
+    val currentTime: Long
+        get() = miniVideo.mediaPlayer.status().time()
+    lateinit var syncOp: (time: Long?) -> Unit
+
+    fun log(message: String): VideoState {
+        if (debug)
+            println("##video## $message")
+        return this
+    }
+
+    fun setPlayerReference(playerState: PlayerState): VideoState {
+        this.playerState = playerState
+        return this
+    }
+
+    fun newestTime(): Long? {
+        return playerState?.framePosition?.let {
+            toMillisecond(it, currentSampleRate).toLong()
+        }
+    }
+
+    fun rememberSize(): VideoState {
+        RememberWidth = width
+        RememberHeight = height
+        return this
+    }
+
+    fun locateVideoPath(samplePath: String): VideoState {
+        videoPath = samplePath.substringBeforeLast('.') + ".mp4"
```

Comment:
We may need some toasts (snackbars) to notify user this path when user tries to show video but no file is found, otherwise, users may get confused how to enable the video feature. We can talk about it later^ ^

Reply:
Maybe using a separate enum class and interface to do video path finding, and catch `FileNotFoundException`:
```suggestion
sealed interface VideoFinder {
    fun findPath(path: String, acceptableExtensions: List<String>): String
}

enum class VideoPath : VideoFinder {
    AtSamePlaceOfReferenceAudio {
        override fun findPath(path: String, acceptableExtensions: List<String>): String {
            val fileName = path.substringBeforeLast('.')
            return acceptableExtensions.asSequence().map{ fileName + it }.filter{File(it).isFile}.firstOrNull() ?: (
                throw FileNotFoundException(
                    "Video not found at the same place of audio ($path)" +
                        "with name $fileName and extensions among $acceptableExtensions"
                )
            )
        }
    },

    UserSpecifyLocation{
        override fun findPath(path: String, acceptableExtensions: List<String>): String {
            TODO("Not yet implemented")
        }
    },
}
```